File: src/lang/Lexer.js
/*
 *  Seni
 *  Copyright (C) 2015 Inderjit Gill <email@indy.io>
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import Token from './Token';
import TokenType from './TokenType';

function characterSet(characters) {
  const s = new Set();

  // todo: is there a better way than iterating over the string?
  for (let i = 0; i < characters.length; i++) {
    s.add(characters[i]);
  }
  return s;
}

const sWhitespaceSet = characterSet(' \t\n,');
const sDigitSet = characterSet('0123456789');
const sAlphaSet =
        characterSet(
          'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ+-*/<>=');
const sSymbolSet = characterSet('-!@#$%^&*<>?');

const MINUS = '-';
const PERIOD = '.';

function isWhitespace(character) {
  return sWhitespaceSet.has(character);
}

function isDigit(character) {
  return sDigitSet.has(character);
}

function isAlpha(character) {
  return sAlphaSet.has(character);
}

function isSymbol(character) {
  return sSymbolSet.has(character);
}

function isListStart(character) {
  return character === '(';
}

function isListEnd(character) {
  return character === ')';
}

function isVectorStart(character) {
  return character === '[';
}

function isVectorEnd(character) {
  return character === ']';
}

function isAlterableStart(character) {
  return character === '{';
}

function isAlterableEnd(character) {
  return character === '}';
}

function isQuotedString(character) {
  return character === '"';
}

function isQuoteAbbreviation(character) {
  return character === '\'';
}

function isComment(character) {
  return character === ';';
}

function isNewline(character) {
  return character === '\n';
}

function isLabel(s) {
  let i = 0;
  for (i = 0; i < s.length; i++) {
    const c = s[i];
    if (!isAlpha(c) && !isDigit(c) && !isSymbol(c)) {
      break;
    }
  }
  return i < s.length && s[i] === ':';
}

// is there a period in the stream of characters before we get to whitespace
function hasPeriod(s) {
  for (let i = 0; i < s.length; i++) {
    if (s[i] === PERIOD) {
      return true;
    }
    if (isWhitespace(s[i])) {
      return false;
    }
  }
  return false;
}

function consumeInt(s) {
  let i = 0;
  for (i = 0; i < s.length; i++) {
    const c = s[i];
    if (!isDigit(c) && c !== MINUS) {
      break;
    }

    if (!isDigit(c) && !(i === 0 && c === MINUS)) {
      break;
    }
  }

  const token = new Token(TokenType.INT, parseInt(s.substring(0, i)));
  return [token, s.substring(i, s.length)];
}

function consumeFloat(s) {
  let i = 0;
  for (i = 0; i < s.length; i++) {
    const c = s[i];
    if (!isDigit(c) && !(i === 0 && c === MINUS) && c !== PERIOD) {
      break;
    }
  }

  const token = new Token(TokenType.FLOAT, parseFloat(s.substring(0, i)));
  return [token, s.substring(i, s.length)];
}

function consumeUnknown(s) {
  return [new Token(TokenType.UNKNOWN, s[0]), s.substring(1)];
}

function consumeListStart(s) {
  return [new Token(TokenType.LIST_START), s.substring(1)];
}

function consumeListEnd(s) {
  return [new Token(TokenType.LIST_END), s.substring(1)];
}

function consumeVectorStart(s) {
  return [new Token(TokenType.VECTOR_START), s.substring(1)];
}

function consumeVectorEnd(s) {
  return [new Token(TokenType.VECTOR_END), s.substring(1)];
}

function consumeAlterableStart(s) {
  return [new Token(TokenType.ALTERABLE_START), s.substring(1)];
}

function consumeAlterableEnd(s) {
  return [new Token(TokenType.ALTERABLE_END), s.substring(1)];
}

function consumeString(s) {
  let val = s.substring(1); // skip first \"
  const nextQuote = val.indexOf('\"');
  val = val.substring(0, nextQuote);

  const token = new Token(TokenType.STRING, val);
  return [token, s.substring(nextQuote + 2)];
}

function consumeName(s) {
  let i = 0;
  for (i = 0; i < s.length; i++) {
    const c = s[i];
    if (!isAlpha(c) && !isDigit(c) && !isSymbol(c)) {
      break;
    }
  }
  const token = new Token(TokenType.NAME, s.substring(0, i));
  return [token, s.substring(i, s.length)];
}


function consumeWhitespace(s) {
  let i = 0;
  for (i = 0; i < s.length; i++) {
    if (!isWhitespace(s[i])) {
      break;
    }
  }
  const token = new Token(TokenType.WHITESPACE, s.substring(0, i));
  return [token, s.substring(i, s.length)];
}

function consumeComment(s) {
  let i = 0;
  for (i = 0; i < s.length; i++) {
    if (isNewline(s[i])) {
      break;
    }
  }
  const token = new Token(TokenType.COMMENT, s.substring(0, i));
  // skip past the newline
  return [token, s.substring(i + 1, s.length)];
}

function consumeLabel(s) {
  let i = 0;
  for (i = 0; i < s.length; i++) {
    const c = s[i];
    if (!isAlpha(c) && !isDigit(c) && !isSymbol(c)) {
      break;
    }
  }
  // read the label name - the ':' character
  const token = new Token(TokenType.LABEL, s.substring(0, i));
  // the remainder should skip past the ':'
  return [token, s.substring(i + 1, s.length)];
}

function consumeQuoteAbbreviation(s) {
  return [new Token(TokenType.QUOTE_ABBREVIATION), s.substring(1)];
}

function nextTokenType(s) {
  const c = s[0];

  if (isWhitespace(c)) {
    return TokenType.WHITESPACE;
  }

  if (isQuoteAbbreviation(c)) {
    return TokenType.QUOTE_ABBREVIATION;
  }

  if (isListStart(c)) {
    return TokenType.LIST_START;
  }

  if (isListEnd(c)) {
    return TokenType.LIST_END;
  }

  if (isVectorStart(c)) {
    return TokenType.VECTOR_START;
  }

  if (isVectorEnd(c)) {
    return TokenType.VECTOR_END;
  }

  if (isAlterableStart(c)) {
    return TokenType.ALTERABLE_START;
  }

  if (isAlterableEnd(c)) {
    return TokenType.ALTERABLE_END;
  }

  if (isQuotedString(c)) {
    return TokenType.STRING;
  }

  if (isAlpha(c)) {
    if (!(c === MINUS && s.length > 1 && isDigit(s[1]))) {
      return isLabel(s) ? TokenType.LABEL : TokenType.NAME;
    }
  }

  if (isDigit(c) || c === MINUS || c === PERIOD) {
    return hasPeriod(s) ? TokenType.FLOAT : TokenType.INT;
  }

  if (isComment(c)) {
    return TokenType.COMMENT;
  }

  return TokenType.UNKNOWN;
}

const Lexer = {
  tokenise: input => {
    const q = [];   // queue of tokens to return
    let p = [];   // [token, remaining] pair

    let s = input;

    while (s.length > 0) {
      switch (nextTokenType(s)) {
      case TokenType.WHITESPACE :
        p = consumeWhitespace(s);
        break;
      case TokenType.LIST_START :
        p = consumeListStart(s);
        break;
      case TokenType.LIST_END :
        p = consumeListEnd(s);
        break;
      case TokenType.VECTOR_START :
        p = consumeVectorStart(s);
        break;
      case TokenType.VECTOR_END :
        p = consumeVectorEnd(s);
        break;
      case TokenType.ALTERABLE_START :
        p = consumeAlterableStart(s);
        break;
      case TokenType.ALTERABLE_END :
        p = consumeAlterableEnd(s);
        break;
      case TokenType.STRING :
        p = consumeString(s);
        break;
      case TokenType.NAME :
        p = consumeName(s);
        break;
      case TokenType.LABEL :
        p = consumeLabel(s);
        break;
      case TokenType.INT :
        p = consumeInt(s);
        break;
      case TokenType.FLOAT :
        p = consumeFloat(s);
        break;
      case TokenType.QUOTE_ABBREVIATION :
        p = consumeQuoteAbbreviation(s);
        break;
      case TokenType.COMMENT :
        p = consumeComment(s);
        break;
      default:
        // read the unknown token and return it
        const tok = consumeUnknown(s)[0];
        return {error: `unknown token: ${tok.value}`,
                tokens: [tok]};
      }

      const [token, remaining] = p;

      q.push(token);
      s = remaining;
    }

    return {tokens: q};
  }
};

export default Lexer;

File: core/sen-core/src/lexer.rs
// Copyright (C) 2018 Inderjit Gill

// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

use error::{SenResult, SenError};

#[derive(Copy, Clone, Debug, Eq, PartialEq)]
pub enum Token<'a> {
    BackQuote,
    Colon,
    Comment(&'a str),
    CurlyBracketEnd,
    CurlyBracketStart,
    DoubleQuote,
    Name(&'a str),
    Number(&'a str),
    ParenEnd,
    ParenStart,
    Quote,
    SquareBracketEnd,
    SquareBracketStart,
    Whitespace(&'a str),
    End,
}

pub fn tokenize(s: &str) -> SenResult<Vec<Token>> {
    let mut lex = Lexer::new(s);
    let mut res = Vec::new();

    loop {
        match lex.eat_token()? {
            Token::End => break,
            tok => res.push(tok),
        }
    }

    Ok(res)
}

struct Lexer<'a> {
    input: &'a str,
    cur_pos: u32,
}

impl<'a> Lexer<'a> {
    pub fn new(input: &str) -> Lexer {
        Lexer{
            input: input,
            cur_pos: 0,
        }
    }

    pub fn eat_token(&mut self) -> SenResult<Token<'a>> {
        let mut chars = self.input.char_indices();

        while let Some((ind, ch)) = chars.next() {
            let res = match ch {
                '(' => Ok((Token::ParenStart, 1)),
                ')' => Ok((Token::ParenEnd, 1)),
                '[' => Ok((Token::SquareBracketStart, 1)),
                ']' => Ok((Token::SquareBracketEnd, 1)),
                '{' => Ok((Token::CurlyBracketStart, 1)),
                '}' => Ok((Token::CurlyBracketEnd, 1)),
                ':' => Ok((Token::Colon, 1)),
                '\'' => Ok((Token::Quote, 1)),
                '"' => Ok((Token::DoubleQuote, 1)),
                '`' => Ok((Token::BackQuote, 1)),
                ';' => eat_comment(&self.input[ind..]),
                '-' | '0' ... '9' => eat_number(&self.input[ind..]),
                ch if ch.is_whitespace() => eat_whitespace(&self.input[ind..]),
                _ if is_name(ch) => eat_name(&self.input[ind..]),
                ch => Err(SenError::ParserInvalidChar(ch))
            };

            let (tok, size) = match res {
                Ok(v) => v,
                Err(kind) => return Err(kind)
            };

            self.cur_pos += size as u32;
            self.input = &self.input[ind + size..];

            return Ok(tok);
        }

        self.input = &self.input[..0];
        Ok(Token::End)
    }
}

fn is_name(ch: char) -> bool {
    ch.is_alphanumeric() || is_symbol(ch)
}

fn is_symbol(ch: char) -> bool {
    match ch {
        '+' | '-' | '*' | '/' | '=' | '!' | '@' |
        '#' | '$' | '%' | '^' | '&' | '<' | '>' |
        '?' => true,
        _ => false
    }
}

fn eat_name(input: &str) -> SenResult<(Token, usize)> {
    for (ind, ch) in input.char_indices() {
        if !is_name(ch) {
            return Ok((Token::Name(&input[..ind]), ind));
        }
    }

    Ok((Token::Name(input), input.len()))
}

fn eat_whitespace(input: &str) -> SenResult<(Token, usize)> {
    for (ind, ch) in input.char_indices() {
        if !ch.is_whitespace() {
            return Ok((Token::Whitespace(&input[..ind]), ind));
        }
    }

    Ok((Token::Whitespace(input), input.len()))
}

fn eat_number(input: &str) -> SenResult<(Token, usize)> {
    let mut digits = false;
    let mut dot = false;
    let mut size = input.len();

    let (prefix_offset, rest) = if input.starts_with('-') {
        match input[1..].chars().next() {
            Some(ch) if ch.is_digit(10) => (1, &input[1..]),
            // Actually a name beginning with '-' rather a number
            _ => return eat_name(input)
        }
    } else {
        (0, input)
    };

    for (ind, ch) in rest.char_indices() {
        match ch {
            '.' => {
                if dot {
                    return Err(SenError::ParserInvalidLiteral);
                }
                dot = true;
            }
            _ if ch.is_digit(10) => {
                digits = true;
            }
            _ => {
                size = prefix_offset + ind;
                break;
            }
        }
    }

    if !digits {
        Err(SenError::ParserInvalidLiteral)
    } else {
        Ok((Token::Number(&input[..size]), size))
    }
}

fn eat_comment(input: &str) -> SenResult<(Token, usize)> {
    let rest = &input[1..];     // remove the first character (;)
    let mut size = rest.len();

    for (ind, ch) in rest.char_indices() {
        match ch {
            '\n' =>{
                size = ind;
                break;
            }
            _ => {
                continue;
            }
        }
    }
    Ok((Token::Comment(&rest[..size]), size + 1))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_lexer() {
        assert_eq!(tokenize("()").unwrap(),
                   [Token::ParenStart,
                    Token::ParenEnd]);

        assert_eq!(tokenize("( )").unwrap(),
                   [Token::ParenStart,
                    Token::Whitespace(" "),
                    Token::ParenEnd]);

        assert_eq!(tokenize("[]").unwrap(),
                   [Token::SquareBracketStart,
                    Token::SquareBracketEnd]);

        assert_eq!(tokenize("{}").unwrap(),
                   [Token::CurlyBracketStart,
                    Token::CurlyBracketEnd]);

        assert_eq!(tokenize("'(1)").unwrap(),
                   [Token::Quote,
                    Token::ParenStart,
                    Token::Number("1"),
                    Token::ParenEnd]);

        assert_eq!(tokenize("5").unwrap(),
                   [Token::Number("5")]);
        assert_eq!(tokenize("-3").unwrap(),
                   [Token::Number("-3")]);
        assert_eq!(tokenize("3.14").unwrap(),
                   [Token::Number("3.14")]);
        assert_eq!(tokenize("-0.34").unwrap(),
                   [Token::Number("-0.34")]);

        assert_eq!(tokenize("1 foo 3").unwrap(),
                   [Token::Number("1"),
                    Token::Whitespace(" "),
                    Token::Name("foo"),
                    Token::Whitespace(" "),
                    Token::Number("3")]);

        assert_eq!(tokenize("hello\nworld").unwrap(),
                   [Token::Name("hello"),
                    Token::Whitespace("\n"),
                    Token::Name("world")]);

        assert_eq!(tokenize("hello ; some comment").unwrap(),
                   [Token::Name("hello"),
                    Token::Whitespace(" "),
                    Token::Comment(" some comment")]);

        assert_eq!(tokenize("hello ; some comment\n(valid)").unwrap(),
                   [Token::Name("hello"),
                    Token::Whitespace(" "),
                    Token::Comment(" some comment"),
                    Token::Whitespace("\n"),
                    Token::ParenStart,
                    Token::Name("valid"),
                    Token::ParenEnd]);

        assert_eq!(tokenize("{ 23 (gen/scalar min: 3 max: 100)}").unwrap(),
                   [Token::CurlyBracketStart,
                    Token::Whitespace(" "),
                    Token::Number("23"),
                    Token::Whitespace(" "),
                    Token::ParenStart,
                    Token::Name("gen/scalar"),
                    Token::Whitespace(" "),
                    Token::Name("min"),
                    Token::Colon,
                    Token::Whitespace(" "),
                    Token::Number("3"),
                    Token::Whitespace(" "),
                    Token::Name("max"),
                    Token::Colon,
                    Token::Whitespace(" "),
                    Token::Number("100"),
                    Token::ParenEnd,
                    Token::CurlyBracketEnd]);
    }
}
